flexmeasures/data/services/sensors.py

```diff
@@ -44,3 +52,109 @@ def get_sensors(
     if sensor_name_allowlist:
         sensor_query = sensor_query.filter(Sensor.name.in_(sensor_name_allowlist))
     return db.session.scalars(sensor_query).all()
+
+
+def get_most_recent_knowledge_time(
+    sensor: Sensor, staleness_search: dict
+) -> datetime | None:
+    """Get the knowledge time of the sensor's most recent event.
+
+    This knowledge time represents when you could have known about the event
+    (specifically, when you could have formed an ex-post belief about it).
+    """
+    staleness_bdf = TimedBelief.search(
+        sensors=sensor,
+        most_recent_events_only=True,
+        **staleness_search,
+    )
+    if staleness_bdf.empty:
+        return None
+    return staleness_bdf.knowledge_times[-1]
+
+
+def get_staleness(
+    sensor: Sensor, staleness_search: dict, now: datetime, is_forecast: bool=False
+) -> timedelta | None:
+    """Get the staleness of the sensor.
+
+    The staleness is defined relative to the knowledge time of the most recent event, rather than to its belief time.
+    Basically, that means that we don't really care when the data arrived,
+    as long as the available data is about what we should be able to know by now.
+
+    :param sensor:              The sensor to compute the staleness for.
+    :param staleness_search:    Deserialized keyword arguments to `TimedBelief.search`.
+    :param now:                 Datetime representing now, used both to mask future beliefs,
+                                and to measures staleness against.
+    :param is_forecast:         Whether the sensor is a forecast sensor.
+    """
+
+    # Mask beliefs before now
+    staleness_search = staleness_search.copy()  # no inplace operations
+    beliefs_before = staleness_search.get("beliefs_before")
+    if beliefs_before is not None:
+        staleness_search["beliefs_before"] = min(beliefs_before, now)
+    elif is_forecast:
+        staleness_search["beliefs_after"] = now
```

Comment:
I don't understand the rationale here. This tries to select beliefs that don't exist yet, but why?

Reply:
We decided to consider sensor status ok when it has data for 12 hours in the future. This way we will look for data with beliefs after now

Reply:
I don't think this is doing what you think it's doing. The `beliefs_after` filters out belief times rather than events.